### PR TITLE
Only use pooled buffers

### PR DESCRIPTION
**Pull request by benaadams**
Only use pooled blocks.
Don't finalize blocks outside debug.

For the cost of just including a finalizer on allocation see:  
http://www.philosophicalgeek.com/2014/09/29/digging-into-net-object-allocation-fundamentals/

### CODE REVIEW

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolBlock2.cs [hunk 1]
```diff
@@ -74,6 +74,9 @@ protected MemoryPoolBlock2()
         /// </summary>
         public MemoryPoolBlock2 Next { get; set; }
 
+#if DEBUG
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolBlock2.cs]: Maybe we should have a Debug.Assert for when `Slab != null` to ensure that we are always returning the block (in our tests at least).

Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolBlock2.cs]: Also probably shouldn't add blocks that aren't pooled to the pool in `Return`

**Comment by benaadams** [comment]
Have something better

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolBlock2.cs (hunk 1 above)
Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolBlock2.cs]: Removed non-pooled blocks instead (from `Lease`)

**Comment by benaadams** [comment]
Revisit this later

**Comment by benaadams** [comment]
Needs some revitalization...

**Pull request closed without merging**